types/src/transaction/mod.rs

```diff
@@ -936,15 +930,47 @@ impl TransactionListWithProof {
             Self::display_option_version(first_transaction_version),
         );
 
-        verify_transaction_list(ledger_info, self)
+        let txn_hashes = self
+            .transactions
+            .iter()
+            .map(|txn| Ok(txn.as_signed_user_txn()?.hash()))
+            .collect::<Result<Vec<_>>>()?;
+        self.proof
+            .verify(ledger_info, self.first_transaction_version, &txn_hashes)?;
+
+        // Verify the events if they exist.
+        if let Some(event_lists) = &self.events {
+            ensure!(
+                event_lists.len() == self.transactions.len(),
+                "The length of event_lists ({}) does not match the number of transactions ({}).",
+                event_lists.len(),
+                self.transactions.len(),
+            );
+            itertools::zip_eq(event_lists, self.proof.transaction_infos())
```

Comment:
nit: `zip_eq` not necessary since we've already checked above that the lens match.

Reply:
I was intentionally using `zip_eq` to be more strict because we've checked above... It panics if the lengths aren't equal.